### ISSUE TITLE
Error reading Clippy's configuration file...

### ISSUE THREAD

**Issue by Diggsey** [rust-lang/rust-clippy#3874]
...  `/opt/.cargo/registry/src/github.com-1ecc6299db9ec823/im-12.3.1/clippy.toml`: found deprecated field `cyclomatic-complexity-threshold`. Please use `cognitive-complexity-threshold` instead.

I use clippy in CI, and started getting this error a few days ago. The error is coming from one of my dependencies, so I don't expect clippy to be checking it to begin with.

Is there a way to stop clippy from checking dependencies / make the error a warning instead?

**Comment by flip1995** [comment]
Yeah errors of clippy.toml files of dependencies should not be reported at all.

**Comment by oli-obk** [comment]
I just checked. Our code for detecting the final binary doesn't seem to work anymore. All dependencies are run through clippy-driver, not sure when that changed. What we can do in the short term is to change the error to a normal lint so it will get silenced by `--cap-lints`.

We should make all the `span_err` or `struct_err` calls in https://github.com/rust-lang/rust-clippy/blob/1cdac4a9c798f8e22a4db220020713175eb9159b/clippy_lints/src/lib.rs#L311 forward to `buffer_lint` (https://doc.rust-lang.org/nightly/nightly-rustc/rustc/session/struct.Session.html#method.buffer_lint).

**Comment by Manishearth** [comment]
The driver stuff got completely rewritten recently since rustc changed how drivers work.

**Comment by JeanMertz** [comment]
I ran into this as well, with a dependency having a `clippy.toml` file with a deprecated field (`cyclomatic-complexity-threshold`).

**Comment by Nemo157** [comment]
I just ran into this as well because of using a very old pinned version of clippy on one project. Quick repro with an invalid config file, might be a different error path to having invalid keys in there:

```console
> cd $(mktemp -d)                                                                                                                                                                                
> cargo new --lib foo
     Created library `foo` package
> cd foo
> cargo add --git https://gist.github.com/Nemo157/963cf5ae958aabb40dbf35dea7b8a9e2 bar
      Adding bar (unknown version) to dependencies
> cargo +stable clippy --all-targets
    Updating git repository `https://gist.github.com/Nemo157/963cf5ae958aabb40dbf35dea7b8a9e2`
    Checking bar v0.1.0 (https://gist.github.com/Nemo157/963cf5ae958aabb40dbf35dea7b8a9e2#b81c54a5)
error: error reading Clippy's configuration file `/home/wim/.cargo/git/checkouts/963cf5ae958aabb40dbf35dea7b8a9e2-9bdfc6a77ca53f42/b81c54a/clippy.toml`: expected an equals, found an identifier at line 1
```

**Comment by izik1** [comment]
I have also ran into this unfortunate bug due to using an unmaintained project as a dependency.

I can take this under the assumption that it's still a `good-first-issue` (presumably if nobody has fixed it, it either isn't as easy as it looks, or is very rare)

**Comment by tylerhawkes** [comment]
I just ran into this with clippy on rust 1.49 for the `pin-project-lite` and `cxx` dependencies. Did something break again?

```
Checking pin-project-lite v0.2.1
error: error reading Clippy's configuration file `/rust/cargo/registry/src/github.com-1ecc6299db9ec823/pin-project-lite-0.2.1/.clippy.toml`: unknown field `msrv`, expected one of `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `pass-by-value-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `vec-box-size-threshold`, `max-trait-bounds`, `max-struct-bools`, `max-fn-params-bools`, `warn-on-all-wildcard-imports`, `disallowed-methods`, `third-party` at line 1 column 1
```

**Comment by taiki-e** [comment]
@tylerhawkes 
I couldn't reproduce the problem you reported in 1.49. 
I guess a version of cargo older than 1.41 as the default is actually called (like https://github.com/tokio-rs/tokio/issues/3355#issuecomment-751482243).
Could you try with toolchain override shorthand (`cargo +stable build`)?

**Comment by tylerhawkes** [comment]
I tried it again and still get the same error. Here's the commands I did to reproduce. I'm inside of a podman container, but that shouldn't matter at all. I tried it on my mac and didn't get an error, so this seems to be linux specific. Cross compiling doesn't matter as I get the error whether I'm compiling for x86_64 or aarch64. 
```
cd /tmp
cargo init clippy-msrv
cd clippy-msrv
echo 'cxx = "1.0.27"' >> Cargo.toml
echo 'pin-project-lite = "0.2.1"' >> Cargo.toml
cat Cargo.toml
# Clean out all of the unzipped crates so they have to be unzipped again
cargo cache -a
CARGO_TARGET_DIR=. cargo +1.49.0-x86_64-unknown-linux-gnu clippy
CARGO_TARGET_DIR=. cargo +1.49.0-x86_64-unknown-linux-gnu clippy
CARGO_TARGET_DIR=. cargo clippy
```

And here is the output of those commands:
```
tmp> cargo init clippy-msrv
     Created binary (application) package
tmp> cd clippy-msrv
clippy-msrv [master ?]> echo 'cxx = "1.0.27"' >> Cargo.toml
clippy-msrv [master ?]> echo 'pin-project-lite = "0.2.1"' >> Cargo.toml
clippy-msrv [master ?]> cat Cargo.toml
[package]
name = "clippy-msrv"
version = "0.1.0"
authors = ["Tyler Hawkes <tyler.hawkes@vivint.com>"]
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
cxx = "1.0.27"
pin-project-lite = "0.2.1"
clippy-msrv [master ?]> cargo cache -a
Cargo cache '/rust/cargo':

Total:                            958.74 MB
  25 installed binaries:          231.15 MB
  Registry:                       458.22 MB
    3 registry indices:           155.09 MB
    752 crate archives:           103.97 MB
    360 crate source checkouts:   199.16 MB
  Git db:                         269.37 MB
    7 bare git repos:              80.83 MB
    3 git repo checkouts:         188.54 MB

Clearing cache...

Size changed from 958.74 MB to 571.04 MB (-387.70 MB, -40.43%)
clippy-msrv [master ?]> CARGO_TARGET_DIR=. cargo +1.49.0-x86_64-unknown-linux-gnu clippy
    Updating crates.io index
   Compiling cc v1.0.66
   Compiling proc-macro2 v1.0.24
   Compiling unicode-xid v0.2.1
   Compiling syn v1.0.58
   Compiling cxxbridge-flags v1.0.27
    Checking pin-project-lite v0.2.1
error: error reading Clippy's configuration file `/rust/cargo/registry/src/github.com-1ecc6299db9ec823/pin-project-lite-0.2.1/.clippy.toml`: unknown field `msrv`, expected one of `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `pass-by-value-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `vec-box-size-threshold`, `max-trait-bounds`, `max-struct-bools`, `max-fn-params-bools`, `warn-on-all-wildcard-imports`, `disallowed-methods`, `third-party` at line 1 column 1

error: aborting due to previous error

error: could not compile `pin-project-lite`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
clippy-msrv [master ?] 101> CARGO_TARGET_DIR=. cargo +1.49.0-x86_64-unknown-linux-gnu clippy
    Checking pin-project-lite v0.2.1
   Compiling link-cplusplus v1.0.4
   Compiling proc-macro2 v1.0.24
   Compiling syn v1.0.58
   Compiling cxx v1.0.27
error: error reading Clippy's configuration file `/rust/cargo/registry/src/github.com-1ecc6299db9ec823/cxx-1.0.27/.clippy.toml`: unknown field `msrv`, expected one of `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `pass-by-value-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `vec-box-size-threshold`, `max-trait-bounds`, `max-struct-bools`, `max-fn-params-bools`, `warn-on-all-wildcard-imports`, `disallowed-methods`, `third-party` at line 1 column 1

error: error reading Clippy's configuration file `/rust/cargo/registry/src/github.com-1ecc6299db9ec823/pin-project-lite-0.2.1/.clippy.toml`: unknown field `msrv`, expected one of `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `pass-by-value-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `vec-box-size-threshold`, `max-trait-bounds`, `max-struct-bools`, `max-fn-params-bools`, `warn-on-all-wildcard-imports`, `disallowed-methods`, `third-party` at line 1 column 1

error: aborting due to previous error

error: could not compile `pin-project-lite`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: aborting due to previous error

error: build failed
clippy-msrv [master ?] 101> CARGO_TARGET_DIR=. cargo clippy
    Checking pin-project-lite v0.2.1
   Compiling proc-macro2 v1.0.24
   Compiling link-cplusplus v1.0.4
   Compiling cxx v1.0.27
error: error reading Clippy's configuration file `/rust/cargo/registry/src/github.com-1ecc6299db9ec823/cxx-1.0.27/.clippy.toml`: unknown field `msrv`, expected one of `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `pass-by-value-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `vec-box-size-threshold`, `max-trait-bounds`, `max-struct-bools`, `max-fn-params-bools`, `warn-on-all-wildcard-imports`, `disallowed-methods`, `third-party` at line 1 column 1

error: error reading Clippy's configuration file `/rust/cargo/registry/src/github.com-1ecc6299db9ec823/pin-project-lite-0.2.1/.clippy.toml`: unknown field `msrv`, expected one of `blacklisted-names`, `cognitive-complexity-threshold`, `cyclomatic-complexity-threshold`, `doc-valid-idents`, `too-many-arguments-threshold`, `type-complexity-threshold`, `single-char-binding-names-threshold`, `too-large-for-stack`, `enum-variant-name-threshold`, `enum-variant-size-threshold`, `verbose-bit-mask-threshold`, `literal-representation-threshold`, `trivial-copy-size-limit`, `pass-by-value-size-limit`, `too-many-lines-threshold`, `array-size-threshold`, `vec-box-size-threshold`, `max-trait-bounds`, `max-struct-bools`, `max-fn-params-bools`, `warn-on-all-wildcard-imports`, `disallowed-methods`, `third-party` at line 1 column 1

error: aborting due to previous error

error: could not compile `pin-project-lite`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: aborting due to previous error

error: build failed
```

**Comment by tylerhawkes** [comment]
My co-workers have also been seeing this. We all share the same podman image and it has `CLIPPY_TESTS=true`. If I set that environment variable on my mac it also has the same issue.
So after a little googling it looks like that got set because we use sccache and I was toying around with getting it to work with clippy. However, [this](https://internals.rust-lang.org/t/clippy-is-available-as-a-rustup-component/7967/26) comment makes it look like clippy is really trying to lint all dependencies and it dies trying to parse the clippy.toml file. It would be nice if stable clippy didn't just die since other crates can include settings that are specific to nightly.

**Comment by flip1995** [comment]
Ah, you have ``CLIPPY_TESTS`` as an env var in your CI/testsuite/...? If that is set to ``true`` (and only ``true``, no other value), then Clippy will lint all dependencies, because it is enabled unconditionally. We need this for Clippys testsuite. 

Since ``CLIPPY_TESTS`` is a reasonable env var for enabling/disabling Clippy tests, we should change the env var, we use internally. This can be done here: https://github.com/rust-lang/rust-clippy/blob/8e5c5a64a5e6689ef25809d9b77380c912572060/src/driver.rs#L301 and in every other place in the Clippy repo, where this var is used. If no one does it until tomorrow, I'll get to it myself.

---

I now wonder, why we never had this complaint before 🤔
